lib/_http_agent.js

```diff
@@ -127,7 +129,16 @@ function Agent(options) {
     const requests = this.requests[name];
     if (requests && requests.length) {
       const req = requests.shift();
-      setRequestSocket(this, req, socket);
+      const reqAsyncRes = req[kRequestAsyncResource];
+      if (reqAsyncRes) {
+        // Run request within the original async context.
+        reqAsyncRes.runInAsyncScope(() => {
+          asyncResetHandle(socket);
+          setRequestSocket(this, req, socket);
+        });
+      } else {
+        setRequestSocket(this, req, socket);
+      }
       if (requests.length === 0) {
```

Comment:
I think we should add `req[kRequestAsyncResource] = null` or similar somewhere here (most likely also for the else case of this section) to allow gc to collect the `AsyncResource` as the socket may stay alive forever.

Reply:
The `if (requests.length === 0)` branch stands for the case when there are no requests left in the queue. So, there is nothing to leak due to the `socket` staying reachable in this case.

Maybe you refer to the `if (reqAsyncRes)`? I've added the `req[kRequestAsyncResource] = null` cleanup into that branch as it will help to GC the `AsyncResource` earlier.

Reply:
The place where you added it is for sure needed and ok.
I was think about queuing happens and a `AsyncResource` is created but the corresponding http request is aborted by user before the `free` event is emitted. In that case we would leak the `AsyncResource`.

Reply:
I see. I'm going to check that scenario and address the comment later today.

Reply:
By looking at the code I can see that `free` event should be emitted with a request is aborted or destroyed: https://github.com/nodejs/node/blob/master/lib/_http_client.js#L382

When this happens, queued requests will be shifted and the destroyed request will be deleted: https://github.com/nodejs/node/blob/master/lib/_http_agent.js#L161

So, I'm not sure if I understand how the leak might happen. Could you provide a bit more details?

Reply:
Right, seems this case is handled in http_client and not visible to http_agent so it should be fine.